### PR TITLE
Implement `openpmd-api` backend as an alternative to `h5py`

### PR DESCRIPTION
**Pull request by RemiLehe**
Action items:
- [x] @ax3l: Implement openPMD-api backend 
- [x] @RemiLehe: Split off the first part of this PR (i.e. isolating `h5py` specific code into a separate folder) into a separate PR: #289
- [x] @ax3l: Rebase this PR once the above-mentioned split-off is done and merged
- [x] Add an installation target so that we can do automatically install the `openpmd-api` backend with `pip install openpmd-viewer[openpmd-api]` (specific syntax is up for discussion :) )
- [ ] add `openpmd_viewer.backend()` function to query active file backend
- [x] remove `openpmd_api` package in one of the matrix entries on travis (so we have continued `h5py` coverage)

Future:
- file-handle management. pro-actively close oldest open files to avoid running into ulimits: https://github.com/openPMD/openPMD-api/pull/822

### CODE REVIEW

File: openpmd_viewer/openpmd_timeseries/data_reader/data_reader.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+                    first_file_name = file_name
+            if first_file_name is None:
+                raise RuntimeError('Unknown files in {0}. '
+                                   'Not found in supported extensions: {1}'
+                                   .format(io.file_extensions))
+
+            # match last occurance of integers and replace with %T wildcards
+            # examples: data00000100.h5 diag4_00000500.h5 io12.0.bp
+            #           te42st.1234.yolo.json scan7_run14_data123.h5
+            file_path = re.sub(r'(\d+)(\.(?!\d).+$)', r'%T\2', first_file_name)
```
Comment (RemiLehe) [inline openpmd_viewer/openpmd_timeseries/data_reader/data_reader.py:79]: 😍

File: openpmd_viewer/openpmd_timeseries/data_reader/data_reader.py [hunk 2]
```diff
@@ -55,6 +64,28 @@ def list_iterations(self, path_to_dir):
                 h5py_reader.list_files( path_to_dir )
             # Store dictionary of correspondence between iteration and file
             self.iteration_to_file = iteration_to_file
+
```
Comment (RemiLehe) [inline openpmd_viewer/openpmd_timeseries/data_reader/data_reader.py:67]: For `h5py`, let us raise a symmetric error to the one below:
```suggestion
            if len(iterations) == 0:
                raise RuntimeError(
                    "Found no valid files in directory {0}.\n"
                    "Please check that this is the path to the openPMD files."
                    "(valid files must have the extension '.h5')"
                    .format(path_to_dir))
```

File: openpmd_viewer/openpmd_timeseries/data_reader/io_reader/field_reader.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+import numpy as np
+from .utilities import get_data
+from openpmd_viewer.openpmd_timeseries.field_metainfo import FieldMetaInformation
+from openpmd_viewer.openpmd_timeseries.utilities import construct_3d_from_circ
+
+
+def read_field_cartesian( series, iteration, field_name, component_name,
+                          axis_labels, slice_relative_position, slice_across ):
+    """
+    Extract a given field from an HDF5 file in the openPMD format,
```
Comment (RemiLehe) [inline openpmd_viewer/openpmd_timeseries/data_reader/io_reader/field_reader.py:20]:
```suggestion
    Extract a given field from a file in the openPMD format,
```

File: openpmd_viewer/openpmd_timeseries/data_reader/io_reader/field_reader.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+            grid_spacing, global_offset,
+            grid_unit_SI, grid_position )
+
+    return F, info 
+
+
+def read_field_circ( series, iteration, field_name, component_name,
+                     slice_relative_position, slice_across, m=0, theta=0. ):
+    """
+    Extract a given field from an HDF5 file in the openPMD format,
```
Comment (RemiLehe) [inline openpmd_viewer/openpmd_timeseries/data_reader/io_reader/field_reader.py:125]:
```suggestion
    Extract a given field from a file in the openPMD format,
```

**Review by RemiLehe — COMMENTED** [review]
Awesome! Thank you so much for the amazing work!
I just a few minor comments, and then this would be ready for merging.

**Comment by ax3l** [comment]
Thank you for the review comments, great points!

**Pull request merged**